scaffolding-node/lib/scaffolding.sh

```diff
@@ -1,277 +1,577 @@
+    declare -g -A scaffolding_process_bins
   fi
+
+  _jq="$(pkg_path_for jq-static)/bin/jq"
 }
 
-# Decide whether or not we need to generate a run hook
-_detect_run_hook() {
-  local start_script_from_package_json run_hook_command
-  start_script_from_package_json=$(_package_json_value pkg_svc_run)
+_detect_package_json() {
+  if [[ ! -f package.json ]]; then
+    exit_with \
+      "Node Scaffolding cannot find package.json in the root directory." 5
+  fi
+  # shellcheck disable=SC2002
+  if ! cat package.json | "$_jq" . > /dev/null; then
+    exit_with "Failed to parse package.json as JSON." 6
+  fi
+
+  # TODO fin: do we check for any lockfiles here?
+}
 
-  build_line "Generating run hook"
-  if [[ -n "$pkg_svc_run" ]]; then
-    build_line "pkg_svc_run=$pkg_svc_run already set; not generating run hook"
-    return 0
-  elif [[ -e "$PLAN_CONTEXT/hooks/run" ]]; then
-    build_line "Existing run hook found in $PLAN_CONTEXT/hooks/run; not generating run hook"
-    return 0
-  elif [[ -n "$start_script_from_package_json" ]]; then
-    build_line "scripts.start set to '$start_script_from_package_json' in package.json; generating run hook using this"
-    run_hook_command=$start_script_from_package_json
-  elif [[ -e "$SRC_PATH/server.js" ]]; then
-    build_line "server.js found in package; generating run hook using this"
-    run_hook_command="node server.js"
+_detect_pkg_manager() {
+  if [[ -n "$scaffolding_pkg_manager" ]]; then
+    case "$scaffolding_pkg_manager" in
+      npm)
+        _pkg_manager=npm
+        build_line "Detected package manager in Plan, using '$_pkg_manager'"
+        ;;
+      yarn)
+        _pkg_manager=yarn
+        build_line "Detected package manager in Plan, using '$_pkg_manager'"
+        ;;
+      *)
+        local e
+        e="Variable 'scaffolding_pkg_manager' can only be"
+        e="$e set to: 'npm', 'yarn', or empty."
+        exit_with "$e" 9
+        ;;
+    esac
+  elif [[ -f yarn.lock ]]; then
+    _pkg_manager=yarn
+    build_line "Detected yarn.lock in root directory, using '$_pkg_manager'"
   else
-    build_line "This package does not look like it runs a service; not generating run hook"
-    return 0
+    _pkg_manager=npm
+    build_line "No package manager could be detected, using default '$_pkg_manager'"
   fi
-
-  _generate_run_hook "$run_hook_command"
 }
 
-# Generate a run hook, given a command
-_generate_run_hook() {
-  local command pkg_name_from_package_json
-  command="$1"
-  # The thing that gets put in lib/node_modules/$name uses the $name from the
-  # package.json, which may be different from the pkg_name
-  pkg_name_from_package_json="$(_package_json_value pkg_name)"
+_detect_process_bins() {
+  local json_start
 
-  mkdir -pv "$pkg_prefix/hooks"
-  cat <<-EOF > "$pkg_prefix/hooks/run"
-#!$(pkg_path_for bash)/bin/bash
-exec 2>&1
-export HOME="$pkg_svc_data_path"
-export NODE_ENV=\${NODE_ENV:-production}
-export NODE_CONFIG_DIR="$pkg_svc_config_path"
```

Comment:
I'm pretty sure https://github.com/lorenwest/node-config is very widely used. How are we expecting node apps to be configured if this is not here? Am I going to have to know about this variable and it's needed value and put it in my scaffolding_env?